In [4]:
import pandas as pd
import numpy as np

df = pd.read_csv("covid19_patient_symptoms_diagnosis.csv")

df_cleaned = df.copy()
df_cleaned['comorbidity'] = df_cleaned['comorbidity'].fillna("None")

symptoms = ['fever', 'dry_cough', 'sore_throat', 'fatigue', 'headache',
            'shortness_of_breath', 'loss_of_smell', 'loss_of_taste', 'chest_pain']

df_cleaned['symptom_count'] = df_cleaned[symptoms].sum(axis=1)

In [5]:
print(df_cleaned.head())

   patient_id  age  gender  fever  dry_cough  sore_throat  fatigue  headache  \
0           1   52    Male      1          0            1        1         0   
1           2   15    Male      0          0            0        1         1   
2           3   72    Male      1          0            1        0         0   
3           4   61  Female      0          0            1        1         1   
4           5   21  Female      1          1            0        0         0   

   shortness_of_breath  loss_of_smell  loss_of_taste  oxygen_level  \
0                    0              0              0            98   
1                    0              0              0            85   
2                    1              0              1            99   
3                    0              1              1            86   
4                    1              0              1            90   

   body_temperature comorbidity  travel_history  contact_with_patient  \
0              37.1    Di

# SECTION 4: DATA MODELING

In this section, we build machine learning models to predict COVID-19 diagnosis.

We use selected features:
- age
- oxygen_level
- body_temperature
- symptom_count

Two models are used:
1. Logistic Regression (baseline model)
2. Random Forest (advanced model)

In [6]:
# select features and target

X = df_cleaned[["age", "oxygen_level", "body_temperature", "symptom_count"]]
y = df_cleaned["covid_result"]

print("Feature shape:", X.shape)
print("Target shape:", y.shape)

Feature shape: (5000, 4)
Target shape: (5000,)


In [7]:
from sklearn.model_selection import train_test_split

# split data (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

print("Training data:", X_train.shape)
print("Testing data:", X_test.shape)

Training data: (4000, 4)
Testing data: (1000, 4)


## Baseline Model: Logistic Regression

In [8]:
from sklearn.linear_model import LogisticRegression

# create and train logistic regression model
log_model = LogisticRegression()
log_model.fit(X_train, y_train)

# make predictions
y_pred_log = log_model.predict(X_test)

print("Logistic Regression model trained successfully.")

Logistic Regression model trained successfully.


In [9]:
from sklearn.metrics import accuracy_score

# calculate accuracy
accuracy_log = accuracy_score(y_test, y_pred_log)

print("Logistic Regression Accuracy:", accuracy_log)

Logistic Regression Accuracy: 0.715


## Advanced Model: Random Forest

In [10]:
from sklearn.ensemble import RandomForestClassifier

# create and train random forest model
rf_model = RandomForestClassifier()
rf_model.fit(X_train, y_train)

# make predictions
y_pred_rf = rf_model.predict(X_test)

print("Random Forest model trained successfully.")

Random Forest model trained successfully.


In [11]:
# calculate accuracy
accuracy_rf = accuracy_score(y_test, y_pred_rf)

print("Random Forest Accuracy:", accuracy_rf)

Random Forest Accuracy: 0.716


## Model Comparison

Logistic Regression Accuracy: 0.715  
Random Forest Accuracy: 0.716  

Both models perform similarly on this dataset.  
Random Forest has slightly higher accuracy, so it performs a little better.

This shows that patient features like age, oxygen level, body temperature, and symptom count can help predict COVID-19 diagnosis.

In [14]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred_rf)
print("Confusion Matrix (Random Forest):")
print(cm)

Confusion Matrix (Random Forest):
[[340 136]
 [148 376]]


### Confusion Matrix Interpretation

The confusion matrix shows how the model performed on predictions.

The model correctly predicts many positive and negative cases, but there are some misclassifications.

Overall, the model performs reasonably well but is not perfect.